pkg/pss/pss_test.go

```diff
@@ -0,0 +1,256 @@
+	}
+
+	registeredHandler := pss.GetHandler(testTopic)
+	registeredHandler(trojan.Message{}) // call handler to verify the retrieved func is correct
+
+	if handlerVerifier != 1 {
+		t.Fatalf("unexpected handler retrieved, verifier variable should be 1 but is %d instead", handlerVerifier)
+	}
+
+	// register second handler
+	testHandler = func(m trojan.Message) {
+		handlerVerifier = 2
+	}
+	testTopic = trojan.NewTopic("SECOND_HANDLER")
+	pss.Register(testTopic, testHandler)
+	if len(pss.GetAllHandlers()) != 2 {
+		t.Fatalf("expected pss handlers to contain 2 elements, but its length is %d", len(pss.GetAllHandlers()))
+	}
+
+	registeredHandler = pss.GetHandler(testTopic)
+	registeredHandler(trojan.Message{}) // call handler to verify the retrieved func is correct
+
+	if handlerVerifier != 2 {
+		t.Fatalf("unexpected handler retrieved, verifier variable should be 2 but is %d instead", handlerVerifier)
+	}
+}
+
+// TestDeliver verifies that registering a handler on pss for a given topic and then submitting a trojan chunk with said topic to it
+// results in the execution of the expected handler func
+func TestDeliver(t *testing.T) {
+	testTags := tags.NewTags()
+	pss := pss.NewPss(pss.Options{logging.New(ioutil.Discard, 0), nil, testTags})
+
+	// test message
+	topic := trojan.NewTopic("footopic")
+	payload := []byte("foopayload")
+	msg, err := trojan.NewMessage(topic, payload)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// test chunk
+	target := trojan.Target([]byte{1}) // arbitrary test target
+	targets := trojan.Targets([]trojan.Target{target})
+	c, err := msg.Wrap(targets)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// trojan chunk has its type set through the validator called by the store, so this needs to be simulated
+	c.WithType(swarm.ContentAddressed)
+
+	// create and register handler
+	var tt trojan.Topic // test variable to check handler func was correctly called
+	hndlr := func(m trojan.Message) {
+		tt = m.Topic // copy the message topic to the test variable
+	}
+	pss.Register(topic, hndlr)
+
+	// call pss Deliver on chunk and verify test topic variable value changes
+	pss.Deliver(c)
+	if tt != msg.Topic {
+		t.Fatalf("unexpected result for pss Deliver func, expected test variable to have a value of %v but is %v instead", msg.Topic, tt)
+	}
+}
+
+func createPusher(t *testing.T, addr swarm.Address, pushSyncService pushsync.PushSyncer, mockOpts ...mocktopology.Option) (*tags.Tags, *pusher.Service, *Store) {
+	t.Helper()
+	logger := logging.New(ioutil.Discard, 0)
+	storer, err := localstore.New("", addr.Bytes(), nil, logger)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	mtags := tags.NewTags()
+	pusherStorer := &Store{
+		Storer:    storer,
+		modeSet:   make(map[string]storage.ModeSet),
+		modeSetMu: &sync.Mutex{},
+	}
+	peerSuggester := mocktopology.NewTopologyDriver(mockOpts...)
+
+	pusherService := pusher.New(pusher.Options{Storer: pusherStorer, PushSyncer: pushSyncService, Tagger: mtags, PeerSuggester: peerSuggester, Logger: logger})
```

Comment:
why do you need a fully blown `pusher` here? wouldn't it be easier to mock it? that's the whole point of having interfaces

Reply:
The mock is not created for pushSync and also it was not necessary for my test.
I removed it.
Thanks